### PR TITLE
Get the SQL working, and make it more idempotent

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/roomserver/storage/event_types_table.go [hunk 1]
```diff
@@ -50,14 +50,13 @@ INSERT INTO event_types (event_type_nid, event_type) VALUES
 // In that case the ID will be assigned using the next value from the sequence.
 // We use `RETURNING` to tell postgres to return the assigned ID.
 // But it's possible that the type was added in a query that raced with us.
-// This will result in a conflict on the event_type_unique constraint.
-// We peform a update that does nothing rather that doing nothing at all because
+// This will result in a conflict on the event_type_unique constraint, in this
+// case we do nothing and rely on the caller to
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/event_types_table.go]: to do what?

File: src/github.com/matrix-org/dendrite/roomserver/input/latest_events.go [hunk 2]
```diff
@@ -79,7 +87,7 @@ func doUpdateLatestEvents(
 		return err
 	}
 
-	if !alreadyReferenced {
+	if !alreadyReferenced && !alreadyInLatest {
 		// This event is not referenced by any of the events in the room.
 		// Add it to the latest events
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/input/latest_events.go]: This comment isn't explaining the deal with `alreadyInLatest`. Please explain.

File: src/github.com/matrix-org/dendrite/roomserver/storage/event_state_keys_table.go [hunk 3]
```diff
@@ -31,8 +31,7 @@ INSERT INTO event_state_keys (event_state_key_nid, event_state_key) VALUES
 const insertEventStateKeyNIDSQL = "" +
 	"INSERT INTO event_state_keys (event_state_key) VALUES ($1)" +
 	" ON CONFLICT ON CONSTRAINT event_state_key_unique" +
-	" DO UPDATE SET event_state_key = $1" +
-	" RETURNING (event_state_key_nid)"
+	" DO NOTHING RETURNING (event_state_key_nid)"
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/event_state_keys_table.go]: Why the change? It felt a lot nicer to return the nid if it already existed rather than having to re-SELECT it.

**Comment by Kegsay** [comment]
Otherwise LGTM.

File: src/github.com/matrix-org/dendrite/roomserver/storage/event_types_table.go (hunk 1 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/event_types_table.go]: Done ae9a226b

File: src/github.com/matrix-org/dendrite/roomserver/input/latest_events.go (hunk 2 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/input/latest_events.go]: Done e0cd3d6e

File: src/github.com/matrix-org/dendrite/roomserver/storage/event_state_keys_table.go (hunk 3 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/event_state_keys_table.go]: 8a6491e

**Comment by Kegsay** [comment]
LGTM
